lib/test_state.go

```diff
@@ -14,20 +15,47 @@ type TestPreInitState struct {
 	Registry       *metrics.Registry
 	BuiltinMetrics *metrics.BuiltinMetrics
 	KeyLogger      io.Writer
+	RootTagSet     *metrics.TagSet
 
 	// TODO: replace with logrus.FieldLogger when all of the tests can be fixed
 	Logger *logrus.Logger
 }
 
+// RunTags create a TagSet for run tags branching from the root TagSet.
+func (s *TestPreInitState) RunTags(opts Options) *metrics.TagSet {
+	// FIXME: remove when we have a stable version
+	// so we don't have to fix the tests multiple times
+	if s.RootTagSet == nil {
+		s.RootTagSet = metrics.NewTagSet(nil)
+	}
+
+	tags := s.RootTagSet.BranchOut()
+	if len(opts.RunTags) < 1 {
+		return tags
+	}
```

Comment:
hmm, I noticed a couple of things here that can maybe be improved, though I am not very sure about that :thinking: 
1. `RootTagSet` might be better off in the `Registry`, as I think you originally said it's going to? be :sweat_smile:  `TestPreInitState` is nice, but you kind of have to initialize its properties in every test :disappointed: Whereas if you add it to `metrics.NewRegistry()`, that will be automatic everyhere, and we'll be even more sure that we have just a single root `Node`. As an extra precaution to ensure that, we can make the root atlas `Node` unexported and have a `GetTagsRoot() *metrics.TagSet` method that returns a new `*metrics.TagSet` wrapper object with the same root node. That way we will have a bulletproof way to ensure we always have exactly one single immutable root.
2. Since `RunTags()` doesn't use all of the `lib.Options` properties, it probably doesn't need to have that as its argument, just a `map[string]string`.
3. Connected to both of the previous points, `RunTags()` also doesn't seem like it should be a method of the `TestPreInitState` :thinking: It can just be a method of `metrics.TagSet`, something like `func (tg *TagSet) BranchOutWith(tags map[string]string) *TagSet { ... }`?

Reply:
>func (tg *TagSet) BranchOutWith(tags map[string]string) *TagSet { ... }

:thinking: Are you sure adding this method doesn't lead to a bad memory usage where we allocate maps when they are not a requirement and iterate using the range pseudo-random order?

I think having only `BranchOut` +  iteration with `AddTag` will force us to do it in an optimal way.

Reply:
Maybe an alternative could be to add `BranchOutWith` to the `registry` instead to add it to the `TagSet`.

Reply:
Check the latest version